### PR TITLE
torch.hub security improvement: add new trust_repo parameter

### PR DESCRIPTION
**Pull request by vmoens**
As pointed by #71205, `torch.hub.load` assumes that the user trusts the repo from where the code is gathered and exececuted. We propose a solution to make sure that the user is aware of the security threat that this can represent.
 
**Solution**: Adds a `trust_repo` parameter to the `load`, `list` and `help` functions in torch.hub.
For now, the default `trust_repo=None` warns that, in the future, the user will need to authorize explicitly every repo before downloading it.
Once the repo has been trusted (via `trust_repo=True` or via a command prompt input) it will be added to the list of trusted repositories.

### CODE REVIEW

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:228]: I feel like we should just return without a message here. IIUC this should only ever happen if the user used `trust_repo=False` while the repo is already trusted?

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:186]: Thoughts on *not* including the branch name?

I feel like it's fair to assume that user would trust an entire repo, and not just some branches of it, but happy to hear your thoughts.

If we're keeping `'_'.join([repo_owner, repo_name, normalized_br]` them let's just declare it above because it's also used to defined `repo_dir`.

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:243]: I think this has already been executed above in `_get_cache_or_reload()`

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:256]: nit: no need for underscore in local variables:
```suggestion
    is_trusted = any([repo.startswith(predefined_trusted) for predefined_trusted in PREDEFINED_TRUSTED])
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:261]: Would something like this be simpler?

```py
trusted_repos = _PREDEFINED_TRUSTED
with open(filepath, 'r') as file:
    trusted_repos += file.readlines()
is_trusted = any(repo.startswith(trusted_repo) for trusted_repo in trusted_repos)
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:281]: This seems to be duplicated with the content of `_add_repo_to_trusted_list()`. Perhaps we could simplify the design here by adding a `prompt` parameter to `_add_repo_to_trusted_list()` instead of a `is_trusted` param?

**Review by NicolasHug — COMMENTED** [review]
Thanks @vmoens , I took a brief look but this looks good. I have a few comments below, LMK what you think. We'll also need to add a few tests ideally (although I'm not sure yet how to handle the prompt)

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:186]: yes I agree, `{repo_owner}_{repo_name}` is sufficient

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:228]: yes that's the case.
I thought that if one asks for `trust_repo=False` then answers `yes` to the prompt, then a notification that this was already trusted could be useful. But happy to remove this if there's no use.

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:261]: sure but with a copy of `_PREDEFINED_TRUSTED`:
```
trusted_repos = [repo for repo in _PREDEFINED_TRUSTED]
```
otherwise if one calls `torch.hub.load` twice in the same session the global list will be incremented twice with the same elements.

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py]: got it, can you check that the current version makes sense to you?

**Comment by vmoens** [comment]
> (although I'm not sure yet how to handle the prompt)
I have written 4 tests: 
1) `trust_repo = False` / `response = ''`
2) `trust_repo = False` / `response = 'y'`
3) `trust_repo = "check"` / `response = 'y'`
4) `trust_repo = None`

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:279]: My suggestion to use a `prompt` parameter isn't very elegant after all. We have some paths that are never hit in `_add_repo_to_trusted_list()` (namely prompt=False and is_trusted=False), and it's a bit weird to call a function with a flag while this flag has just been checked before the call (is_trusted).

After all, we only want to prompt in 2 cases:

- if trust_repo=False
- if trust_repo == "check" and not is_trusted


I tried to design the code below around this observation, removing the `_add_repo_to_trusted_list()` function. WDYT? (I haven't checked whether it works or not lol) 


```py
    if (trust_repo is False) or (trust_repo == "check" and not is_trusted):
        response = input(
            f"The repository {repo} does not belong to the list of trusted repositories and as such cannot be downloaded. "
            "Do you trust this repository and wish to add it to the trusted list of repositories (y/N)?"
        )
        if response.lower() in ("y", "yes"):
            if is_trusted:
                print("The repository is already trusted.")
        elif response.lower() in ("n", "no", ""):
            raise Exception("Untrusted repository.")
        else:
            raise ValueError(f"Unrecognized response {response}.")

    # At this point we're sure that the user trusts the repo
    if not is_trusted:
        with open(filepath, "a") as file:
            file.write(repo + "\n")
```

Comment (vmoens) [reply torch/hub.py:279]: I agree that's way clearer than before

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:215]: Could we just patch `input` instead? When we can, it's preferable to not alter the code just because of testing constraints.

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:165]: Declaring a default suggest that the function can be called as `_get_cache_or_reload(github=..., force_reload=...)`, but that's never the case.


```suggestion
def _get_cache_or_reload(github, force_reload, trust_repo, calling_fn, verbose=True, skip_validation=False):
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:222]: Similarly to comment above

```suggestion
def _check_repo(repo, trust_repo, calling_fn):
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:241]: Let's give our future selves a bit more details: 

```suggestion
    # TODO: Remove `None` option in future version and change the default to "check"
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:354]: This does not render line by line, which I think was the original intent (rendered docs: https://docs-preview.pytorch.org/72060/hub.html) . I think we'll need to write it as a list and skip lines where appropriate.
Also made some minor edits


```suggestion
        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
             This parameter helps ensuring that users only run code from repos that they trust.
             
            - If ``False``, a prompt will ask the user whether the repo should be trusted.
            - If ``True``, the repo will be added to the trusted list and loaded without requiring explicit confirmation.
            - If ``"check"``, the repo will be checked against the list of trusted repos in the cache. If it is
               not present in that list, the behaviour will fall back onto the ``trust_repo=False`` option.
            - If ``None``: this will raise a warning, inviting the user to set ``trust_repo`` to either ``False``, ``True`` or ``"check"``. This is only present for backward compatibility and will be removed in the future.
            
            Default is ``None`` and will eventually change to ``"check"`` in a future version.
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:245]:
```suggestion
                "You are about to download and run code from an untrusted repository. In a future release, this won't be allowed. "
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:249]:
```suggestion
                f"or {calling_fn}(..., trust_repo=True), which will assume that the prompt is to be answered with "
                f"'yes'. You can also use {calling_fn}(..., trust_repo="check") which will only prompt for confirmation if the repo is not already trusted. This will eventually be the default behaviour").
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:233]: I think this will add repos of the form `repoowner_reponame_branch` like `pytorch_vision_main`, and these won't be matched against what we actually check which is just `repoowner_reponame`.

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:239]: As discussed offline we'll need to avoid `startswith` to avoid malicious repo owners like `pytorch_malicious/bad_repo`

**Review by NicolasHug — COMMENTED** [review]
Thanks @vmoens , I made another pass, I'll look a the tests later!

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:215]: Nope, I get

```python
TypeError: Need a valid target to patch. You supplied: 'input'
```

Comment (NicolasHug) [reply torch/hub.py:215]: Perhaps we need to specify builtins.input or something like that?

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py]: That should be fixed by now

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:233]: good point, that comes from the previous check with the `owner_repo_banch` strings.
We could split after the last `_` but that won't work when the branch is `my_branch_name`.
Do we keep track of what the owner / repo / branch is for a given cached repo?

Comment (vmoens) [reply torch/hub.py:233]: I don't think we do
What we can do is this: We have 3 sources for the checks 
(1) `_TRUSTED_REPO_OWNERS` tuple that checks the owners only
(2) `trusted_list` file, that checks `owner_repo`
(3) `trusted_list_legacy`  file, that checks `owner_repo_branch` since it's all we have

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:244]: Nit: it might be easier to follow what's going with:

```suggestion
    owner_name = '_'.join([repo_owner, repo_name])
    owner_name_branch = '_'.join([repo_owner, repo_name, repo_branch])
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:459]: we'll have to udpate these docs, and those of `help` too.

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:355]: There are some indentation issues here and below that lead to incorrect rendered docs: https://docs-preview.pytorch.org/72060/hub.html

I think this should do the job:

```py
        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
            This parameter helps ensuring that users only run code from repos that they trust.

            - If ``False``, a prompt will ask the user whether the repo should
              be trusted.
            - If ``True``, the repo will be added to the trusted list and loaded
              without requiring explicit confirmation.
            - If ``"check"``, the repo will be checked against the list of
              trusted repos in the cache. If it is not present in that list, the
              behaviour will fall back onto the ``trust_repo=False`` option.
            - If ``None``: this will raise a warning, inviting the user to set
              ``trust_repo`` to either ``False``, ``True`` or ``"check"``. This
              is only present for backward compatibility and will be removed in
              the future.

            Default is ``None`` and will eventually change to ``"check"`` in a future version.
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:238]:
```suggestion
```

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:240]: is the `if len(line.strip())` part necessary?

If it is, a more Pythonic way is simply `if line.strip()`

File: torch/hub.py
Comment (NicolasHug) [inline torch/hub.py:248]: This may or may not be neater, I'll let you decide lol

```suggestion
    is_trusted = (
        any(repo == trusted_repo for trusted_repo in trusted_repos)
        or any(repo_branch == trusted_repo for trusted_repo in trusted_repos_legacy)
        or repo_owner in _TRUSTED_REPO_OWNERS
)
```

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:762]: Here and in other tests it looks like the `input` parameter isn't used

```suggestion
    def test_untrusted_repo(self):
```

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:780]: Let's also test the case when `trust_repo=True`
I know we test it everywhere else but it's worth having it here too for "documentation"

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:783]: Let's also check if the return value in "no"

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:758]: Do we actually need these files?

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:736]: Maybe we should make this a `TearDown()` method instead? https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:817]: There should be no prompt, right? I think we should be able to remove the mock entirely and the comment could become

> the repo + branch is in the legacy list so it is considered a trusted repo, and the user isn't prompted anything

File: test/test_utils.py
Comment (NicolasHug) [inline test/test_utils.py:815]: We'll have to remove these files at the end of the test so as not to influence other tests

**Review by NicolasHug — COMMENTED** [review]
Thanks @vmoens , I made another pass!

File: torch/hub.py
Comment (vmoens) [reply torch/hub.py:247]: I prefer mine for debugging :p

File: test/test_utils.py
Comment (vmoens) [reply test/test_utils.py:833]: now that we have `tearDown` the written lines should be removed

**Review by NicolasHug — APPROVED** [review]
Thanks a lot @vmoens ,

I added a few tests and also made sure to check how many times `input()` was called, to make sure we prompt the user only when we need to. I also removed the use of the `legacy_file` and instead just check the content of the cache directly, as discussed offline. I'll approve the PR, if you're OK with my own change let's merge this next week and road-test it :) !

**Comment by vmoens** [comment]
> Thanks a lot @vmoens ,
> 
> I added a few tests and also made sure to check how many times `input()` was called, to make sure we prompt the user only when we need to. I also removed the use of the `legacy_file` and instead just check the content of the cache directly, as discussed offline. I'll approve the PR, if you're OK with my own change let's merge this next week and road-test it :) !

Awesome sounds good to me! Thanks for the many improvements

**Review by NicolasHug — COMMENTED** [review]
@pytorchmergebot please merge this

**Pull request closed without merging**
